tensornetwork/matrixproductstates/mps.py

```diff
@@ -95,20 +87,330 @@ def dtype(self):
+      site1: the site where `op1`  acts
+      sites2: sites where `op2` acts.
+    Returns:
+      List: correlator <op1, op2>
+    Raises:
+      ValueError if `site1` is out of range
+    """
+    N = len(self)
+    if site1 < 0:
+      raise ValueError(
+          "Site site1 out of range: {} not between 0 <= site < N = {}.".format(
+              site1, N))
+    sites2 = np.array(sites2)  #enable logical indexing
+
+    # we break the computation into two parts:
+    # first we get all correlators <op2(site2) op1(site1)> with site2 < site1
+    # then all correlators <op1(site1) op2(site2)> with site1 >= site1
+
+    # get all sites smaller than site1
+    left_sites = sorted(sites2[sites2 < site1])
+    # get all sites larger than site1
+    right_sites = sorted(sites2[sites2 > site1])
+
+    # compute all neccessary right reduced
+    # density matrices in one go. This is
+    # more efficient than calling right_envs
+    # for each site individually
+    if right_sites:
+      right_sites_mod = list({n % N for n in right_sites})
+      rs = self.right_envs([site1] + right_sites_mod)
+    c = []
+    if left_sites:
+
+      left_sites_mod = list({n % N for n in left_sites})
+
+      ls = self.left_envs(left_sites_mod + [site1])
+      A = Node(self.nodes[site1], backend=self.backend.name)
+      O1 = Node(op1, backend=self.backend.name)
+      conj_A = conj(A)
+      R = rs[site1]
+      R[0] ^ A[2]
+      R[1] ^ conj_A[2]
+      A[1] ^ O1[1]
+      conj_A[1] ^ O1[0]
+      R = ((R @ A) @ O1) @ conj_A
+      n1 = np.min(left_sites)
+      #          -- A--------
+      #             |        |
+      # compute   op1(site1) |
+      #             |        |
+      #          -- A*-------
+      # and evolve it to the left by contracting tensors at site2 < site1
+      # if site2 is in `sites2`, calculate the observable
+      #
+      #  ---A--........-- A--------
+      # |   |             |        |
+      # |  op2(site2)    op1(site1)|
+      # |   |             |        |
+      #  ---A--........-- A*-------
+
+      for n in range(site1 - 1, n1 - 1, -1):
+        if n in left_sites:
+          A = Node(self.nodes[n % N], backend=self.backend.name)
+          conj_A = conj(A)
+          O2 = Node(op2, backend=self.backend.name)
+          L = ls[n % N]
+          L[0] ^ A[0]
+          L[1] ^ conj_A[0]
+          O2[0] ^ conj_A[1]
+          O2[1] ^ A[1]
+          R[0] ^ A[2]
+          R[1] ^ conj_A[2]
+
+          res = (((L @ A) @ O2) @ conj_A) @ R
+          c.append(res.tensor)
+        if n > n1:
+          R = self.apply_transfer_operator(n % N, 'right', R)
+
+      c = list(reversed(c))
+
+    # compute <op2(site1)op1(site1)>
+    if site1 in sites2:
+      O1 = Node(op1, backend=self.backend.name)
+      O2 = Node(op2, backend=self.backend.name)
+      L = ls[site1]
+      R = rs[site1]
+      A = Node(self.nodes[site1], backend=self.backend.name)
+      conj_A = conj(A)
```

Comment:
It would be nice to have the `with` statement here. #339 (no changes needed at this time).

Reply:
true, I forgot that we have this now. Will add with a next PR!

Reply:
We DON'T have it now actually.